ambry-frontend/src/main/java/com.github.ambry.frontend/AmbryBlobStorageService.java

```diff
@@ -995,4 +1003,208 @@ void markStartTime() {
+      throw new RestServiceException("Account cannot be found for accountName=" + accountName + " in new put request",
+          RestServiceErrorCode.InvalidAccount);
+    }
+    ensureAccountNameMatchOrThrow(targetAccount, restRequest);
+    String containerName = getTargetContainerNameFromHeader(restRequest);
+    Container targetContainer = targetAccount.getContainerByName(containerName);
+    if (targetContainer == null) {
+      frontendMetrics.newPutUnRecognizedContainerCount.inc();
+      throw new RestServiceException(
+          "Container cannot be identified for accountName=" + accountName + " and containerName=" + containerName
+              + " in new put request.", RestServiceErrorCode.InvalidContainer);
+    }
+    ensureContainerNameMatchOrThrow(targetContainer, restRequest);
+    setTargetAccountAndContainerInRestRequest(restRequest, targetAccount, targetContainer);
+  }
+
+  /**
+   * Obtains the target {@link Account} and {@link Container} id from the blobId string, queries the {@link AccountService}
+   * to get the corresponding {@link Account} and {@link Container}, and injects the target {@link Account} and
+   * {@link Container} into the {@link RestRequest}.
+   * @param blobIdStr The blobId string to get the target {@link Account} and {@link Container} id。
+   * @param restRequest The rest request to insert the target {@link Account} and {@link Container}.
+   * @throws RestServiceException if 1) either {@link Account} or {@link Container} could not be identified; or 2)
+   *                              either {@link Account} or {@link Container} were explicitly specified as
+   *                              {@link Account#UNKNOWN_ACCOUNT} or {@link Container#UNKNOWN_CONTAINER}.
+   */
+  private void injectTargetAccountAndContainerFromBlobId(String blobIdStr, RestRequest restRequest)
+      throws RestServiceException {
+    try {
+      BlobId blobId = new BlobId(blobIdStr, clusterMap);
+      Account targetAccount = accountService.getAccountById(blobId.getAccountId());
+      if (targetAccount == null) {
+        frontendMetrics.getHeadDeleteUnrecognizedAccountCount.inc();
+        // @todo The check can be removed once HelixAccountService is running with UNKNOWN_ACCOUNT created.
+        if (blobId.getAccountId() != Account.UNKNOWN_ACCOUNT_ID) {
+          throw new RestServiceException(
+              "Account from blobId=" + blobIdStr + "with accountId=" + blobId.getAccountId() + " cannot be recognized",
+              RestServiceErrorCode.InvalidAccount);
+        } else {
+          logger.warn(
+              "Account cannot be identified for blobId={} with accountId={}. Setting targetAccount to UNKNOWN_ACCOUNT",
+              blobIdStr, blobId.getAccountId());
+          targetAccount = Account.UNKNOWN_ACCOUNT;
+        }
+      }
+      Container targetContainer = targetAccount.getContainerById(blobId.getContainerId());
+      if (targetContainer == null) {
```

Comment:
Will come talk to you regarding how GET requests for old blobs will be handled here.